electrum/plugins/ledger/ledger.py

```diff
@@ -540,75 +718,547 @@ def sign_transaction(self, tx, password):
+                            )
+                        else:
+                            # register the policy
+                            if script_addrtype == AddressType.LEGACY:
+                                name = "Legacy"
+                                template = "pkh(@0/**)"
+                            elif script_addrtype == AddressType.WIT:
+                                name = "Native SegWit"
+                                template = "wpkh(@0/**)"
+                            elif script_addrtype == AddressType.SH_WIT:
+                                name = "Nested SegWit"
+                                template = "sh(wpkh(@0/**))"
+                            elif script_addrtype == AddressType.TAP:
+                                name = "Taproot"
+                                template = "tr(@0/**)"
+                            else:
+                                raise ValueError("Unknown address type")
+
+                            key_origin_info = origin.to_string()
+                            key_origin_steps = key_origin_info.replace('h', '\'').split('/')[1:]
+                            if len(key_origin_steps) < 3:
+                                # Skip this input, not able to sign
+                                return
+
+                            # remove the last two steps
+                            account_key_origin = "/".join(key_origin_steps[:-2])
+
+                            # get the account-level xpub
+                            xpub = self.get_xpub(f"m/{account_key_origin}", 'standard')
+                            key_str = f"[{master_fp.hex()}/{account_key_origin}]{xpub}"
+
+                            policy = WalletPolicy(name, template, [key_str])
+                            __, registered_hmac = self.client.register_wallet(policy)
+                            wallets[policy.id] = (
+                                script_addrtype,
+                                policy,
+                                registered_hmac,
+                            )
+                    for key, origin in psbt_in.hd_keypaths.items():
+                        if origin.fingerprint == master_fp:
+                            process_origin(origin)
+
+                    for key, (__, origin) in psbt_in.tap_bip32_paths.items():
+                        # TODO: Support script path signing
+                        if key == psbt_in.tap_internal_key and origin.fingerprint == master_fp:
+                            process_origin(origin)
+
+            self.handler.show_message(_("Confirm Transaction on your Ledger device..."))
+
+            if len(wallets) == 0:
+                # Could not find a WalletPolicy to sign with
+                raise UserFacingException(_('Unable to sign this transaction'))
+
+            # For each wallet, sign
+            for __, (__, wallet, wallet_hmac) in wallets.items():
+                input_sigs = self.client.sign_psbt(psbt, wallet, wallet_hmac)
```

Comment:
This new API is simpler than the old one but has a significant regression: there is no feedback whatsoever to the user regarding progress. see https://github.com/spesmilo/electrum/issues/7516 and in particular https://github.com/spesmilo/electrum/commit/ec98ef5aee72d5b5e5ec119a1c176820c6bb4635
The legacy client gives feedback to the user via callbacks (shown in the UI). The new client is a blackbox.
Note that unlike e.g. trezor, ledger does not show signing progress on the device screen either.
Signing a 70 input tx takes 5+ minutes, during which there is no feedback anywhere with the new client.

---

I've just tested signing a few txs, all p2wpkh inputs and outputs (see commit linked above re phase1/phase2 distinction):
"nano s", bitcoin app 1.6.3
10 txins
- p1: 49 sec
- p2: 17 sec

"nano s", bitcoin app 1.6.3
70 txins
- p1: 328 sec
- p1: 122 sec

---

"nano s plus", bitcoin app 2.1.0rc3
10 txins
- p1: 50 sec
- p2: 57 sec

"nano s plus", bitcoin app 2.1.0rc3
70 txins
- error after 5 mins:
	```
	E | plugins.ledger.ledger | Error while signing
	Traceback (most recent call last):
	  File ".../electrum/electrum/plugins/ledger/ledger.py", line 1154, in sign_transaction
		input_sigs = self.client.sign_psbt(psbt, wallet, wallet_hmac)
	  File ".../site-packages/ledger_bitcoin/client.py", line 189, in sign_psbt
		raise DeviceException(error_code=sw, ins=BitcoinInsType.SIGN_PSBT)
	ledger_bitcoin.exception.errors.UnknownDeviceError: ('0x5515', 'Error in <BitcoinInsType.SIGN_PSBT: 4> command', '')
	```
        That is, I could not sign the tx using the new client, only using legacy, as the device errored.

---

This is not a blocker for merging, but it would be good to improve on later. Either the device should show progress on-screen (as trezor), or the API could be extended so that we can pass a callback.

Reply:
> In any case, it's weird for the PIN lock to activate mid-signing :P

Agreed, but this might be tricky because of the constraints from the OS (applications being able to subvert the PIN lock might be considered a breach in the security model).

> (btw timings are: p1: 300 sec, p2: 393 sec)

There is some room for improvement, especially on NanoX/NanoS+; adding [wallet policies](https://github.com/LedgerHQ/app-bitcoin-new/blob/develop/doc/wallet.md) added lots of new possibilities for multisig/miniscript, but focus was on making sure it still fits on Nano S rather than optimizing performance (so there's some redundant work).
Hopefully not too many people frequently send transactions with tens of inputs, yet!

Hopefully I'll find time for making a future PR to further improve the UX for multisignature (persisting the wallet registration on disk would be the main missing feature).

Thanks a lot for your time!